compiler/frontend/src/org/jetbrains/kotlin/types/expressions/LabelResolver.kt

```diff
@@ -17,49 +17,64 @@
 import org.jetbrains.kotlin.psi.*
 import org.jetbrains.kotlin.psi.psiUtil.checkReservedYield
 import org.jetbrains.kotlin.resolve.*
-import org.jetbrains.kotlin.resolve.BindingContext.LABEL_TARGET
-import org.jetbrains.kotlin.resolve.BindingContext.REFERENCE_TARGET
+import org.jetbrains.kotlin.resolve.BindingContext.*
 import org.jetbrains.kotlin.resolve.calls.context.ResolutionContext
 import org.jetbrains.kotlin.resolve.scopes.utils.getDeclarationsByLabel
+import org.jetbrains.kotlin.utils.addIfNotNull
 
 object LabelResolver {
-    private fun getElementsByLabelName(labelName: Name, labelExpression: KtSimpleNameExpression): Set<KtElement> {
+    private fun getElementsByLabelName(
+        labelName: Name,
+        labelExpression: KtSimpleNameExpression,
+        classNameLabelsEnabled: Boolean
+    ): Set<KtElement> {
         val elements = linkedSetOf<KtElement>()
         var parent: PsiElement? = labelExpression.parent
         while (parent != null) {
-            val name = getLabelNameIfAny(parent)
-            if (name != null && name == labelName) {
+            val names = getLabelNamesIfAny(parent, classNameLabelsEnabled)
+            if (names.contains(labelName)) {
                 elements.add(getExpressionUnderLabel(parent as KtExpression))
             }
             parent = if (parent is KtCodeFragment) parent.context else parent.parent
         }
         return elements
     }
 
-    fun getLabelNameIfAny(element: PsiElement): Name? {
-        return when (element) {
-            is KtLabeledExpression -> element.getLabelNameAsName()
-            is KtFunctionLiteral -> getLabelNameIfAny(element.parent)
-            is KtLambdaExpression -> getLabelForFunctionalExpression(element)
-            is KtNamedFunction -> element.nameAsName ?: getLabelForFunctionalExpression(element)
-            else -> null
+    fun getLabelNamesIfAny(element: PsiElement, addClassNameLabels: Boolean): List<Name> {
+        val result = mutableListOf<Name>()
+        when (element) {
+            is KtLabeledExpression -> result.addIfNotNull(element.getLabelNameAsName())
+            // TODO: Support context receivers in function literals
+            is KtFunctionLiteral -> return getLabelNamesIfAny(element.parent, false)
+            is KtLambdaExpression -> result.addIfNotNull(getLabelForFunctionalExpression(element))
+        }
+        val functionOrProperty = when (element) {
+            is KtNamedFunction -> element
+            is KtPropertyAccessor -> element.property
+            else -> return result
+        }
+        if (addClassNameLabels) {
+            functionOrProperty.receiverTypeReference?.nameForReceiverLabel()?.let { result.add(Name.identifier(it)) }
+            functionOrProperty.contextReceivers
+                .mapNotNullTo(result) { it.name()?.let { s -> Name.identifier(s) } }
         }
+        val name = functionOrProperty.nameAsName ?: getLabelForFunctionalExpression(functionOrProperty)
```

Comment:
Can you please make sure, we don't add same-named label for properties
Currently, this doesn't work
```

fun myPrint(x: Int) {}
```

Reply:
I didn't quite get what this code snippet shows, but you're right, here I changed the original semantics. Thanks for noticing, fixed in [this commit](https://github.com/JetBrains/kotlin/pull/4488/commits/d29c2b72b9278549146615a6c71f3f82ba23df8a).